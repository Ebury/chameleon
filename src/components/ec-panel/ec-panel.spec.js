import { mount, createLocalVue } from '@vue/test-utils';
import EcPanel from './ec-panel.vue';

function mountPanel(props, mountOpts) {
  return mount(EcPanel, {
    propsData: { ...props },
    ...mountOpts,
  });
}

function mountPanelAsTemplate(template, props, mountOpts) {
  const localVue = createLocalVue();

  const Component = localVue.extend({
    components: { EcPanel },
    template,
  });

  return mount(Component, {
    localVue,
    propsData: { ...props },
    ...mountOpts,
  });
}

describe('EcPanel', () => {
  describe(':props', () => {
    it(':show - should render the panel when is true', () => {
      const wrapper = mountPanel({ show: true });

      expect(wrapper.find('.ec-panel').exists()).toBe(true);
      expect(wrapper.element).toMatchSnapshot();
    });

    it(':show - should not render the panel when is false', () => {
      const wrapper = mountPanel({ show: false });

      expect(wrapper.find('.ec-panel').exists()).toBe(false);
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('@events', () => {
    it('@close - should emit both "show-panel" and "close" events when the simple-close icon is clicked', () => {
      const wrapper = mountPanel({ show: true });
      wrapper.find('.ec-panel__header-action--close').trigger('click');

      expect(wrapper.emitted('show-panel').length).toBe(1);
      expect(wrapper.emitted('close').length).toBe(1);
    });

    describe('@back', () => {
      it('should render back button when event handler attribute is present', () => {
        const wrapper = mountPanelAsTemplate(
          '<ec-panel v-model="show" @back="anyGivenCallback"></ec-panel>',
          {},
          {
            data() {
              return {
                show: true,
              };
            },
            methods: {
              anyGivenCallback: jest.fn(),
            },
          },
        );

        expect(wrapper.find('.ec-panel__header-action--back').exists()).toBeTruthy();
        expect(wrapper.element).toMatchSnapshot();
      });

      it('should not render back button when event handler attribute is not present', () => {
        const wrapper = mountPanelAsTemplate(
          '<ec-panel v-model="show"></ec-panel>',
          {},
          {
            data() {
              return {
                show: true,
              };
            },
          },
        );

        expect(wrapper.find('.ec-panel__header-action--back').exists()).toBeFalsy();
        expect(wrapper.element).toMatchSnapshot();
      });

      it('should emit a "back" event when the simple-chevron-left icon is clicked', () => {
        const anyGivenCallback = jest.fn();
        const wrapper = mountPanelAsTemplate(
          '<ec-panel v-model="show" @back="anyGivenCallback"></ec-panel>',
          {},
          {
            data() {
              return {
                show: true,
              };
            },
            methods: {
              anyGivenCallback,
            },
          },
        );

        wrapper.find('.ec-panel__header-action--back').trigger('click');

        expect(anyGivenCallback).toHaveBeenCalled();
        expect(wrapper.find('.ec-panel').exists()).toBeFalsy();
      });
    });
  });

  describe('#slots', () => {
    it('#header - should render header slot section if slot is passed', () => {
      const wrapper = mountPanel(
        { show: true },
        {
          slots: {
            header: '<h2>Submitted Requests</h2>',
          },
        },
      );
      expect(wrapper.element).toMatchSnapshot();
    });

    it('#main - should render main slot section if slot is passed', () => {
      const wrapper = mountPanel(
        { show: true },
        {
          slots: {
            main: '<p>Main content section</p>',
          },
        },
      );
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('v-model', () => {
    it('should render the panel when we pass to model true', () => {
      const wrapper = mountPanelAsTemplate(
        '<ec-panel v-model="show"></ec-panel>',
        {},
        {
          data() {
            return {
              show: true,
            };
          },
        },
      );

      expect(wrapper.find('.ec-panel').exists()).toBe(true);
      expect(wrapper.element).toMatchSnapshot();

      wrapper.find('.ec-panel__header-action--close').trigger('click');
      expect(wrapper.vm.show).toBe(false);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should not render the panel when we pass to model false', () => {
      const wrapper = mountPanelAsTemplate(
        '<ec-panel v-model="show"></ec-panel>',
        {},
        {
          data() {
            return {
              show: false,
            };
          },
        },
      );

      expect(wrapper.find('.ec-panel').exists()).toBe(false);
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
