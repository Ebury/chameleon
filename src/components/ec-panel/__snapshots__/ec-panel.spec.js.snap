// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcPanel #slots #footer - should render footer slot section if slot is passed 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <div
        class="ec-panel__footer"
        data-test="ec-panel__footer"
      >
        <h2>
          Footer Section
        </h2>
      </div>
    </div>
  </div>
</aside>
`;

exports[`EcPanel #slots #header - should render header slot section if slot is passed 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <div
        class="ec-panel__header"
        data-test="ec-panel__header"
      >
        <h2>
          Header Section
        </h2>
      </div>
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel #slots #main - should render main slot section if slot is passed 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      >
        <p>
          Main content section
        </p>
      </div>
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel :props :show - should not render the panel when is false 1`] = `<!---->`;

exports[`EcPanel :props :show - should render the panel when is true 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel @events @back should not render back button when event handler attribute is not present 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel @events @back should render back button when event handler attribute is present 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <button
        aria-label="Go back"
        class="ec-panel__header-action ec-panel__header-action--back"
        data-test="ec-panel__header-action--back"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-arrow-left"
          />
        </svg>
      </button>
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel v-model should not render the panel when we pass to model false 1`] = `<!---->`;

exports[`EcPanel v-model should render the panel when we pass to model true 1`] = `
<aside
  class="ec-panel"
  data-test="ec-panel"
>
  <div
    class="ec-panel__fixed-container"
    data-test="ec-panel__fixed-container"
  >
    <div
      class="ec-panel__header-icons"
      data-test="ec-panel__header-icons"
    >
      <!---->
       
      <button
        aria-label="Close panel"
        class="ec-panel__header-action ec-panel__header-action--close"
        data-test="ec-panel__header-action--close"
        type="button"
      >
        <svg
          class="ec-panel__header-icon ec-icon"
          height="24"
          width="24"
        >
          <use
            href="/img/simple-icons.svg#ec-simple-close"
          />
        </svg>
      </button>
    </div>
     
    <div
      class="ec-panel__content-container"
    >
      <!---->
       
      <div
        class="ec-panel__main"
        data-test="ec-panel__main"
      />
       
      <!---->
    </div>
  </div>
</aside>
`;

exports[`EcPanel v-model should render the panel when we pass to model true 2`] = `<!---->`;
